keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,309 @@
+
+```yaml
+cosi.io/provisioner-secret-name:
+cosi.io/provisioner-secret-namespace:
+cosi.io/user-secret-name:
+cosi.io/user-secret-namespace:
+```
+
+Bucket automation should support templating of Secret names and namespaces. For example:
+
+```yaml
+cosi.io/user-secret-name: "${bucket.name}"
+cosi.io/user-secret-namespace: "${bucket.namespace}-region-west"
+```
+
+> Note: Annotation key templates are not mentioned as of yet.  However, need for them may arise during automation design.
+
+## Custom Resource Definitions
+
+#### Bucket
+
+A user facing API object representing an object store bucket. Created by a user in their app's namespace. Once provisiong is complete, the `Bucket` is "bound" to the corresponding `BucketContent`. Binding is 1:1, meaning there is only one `BucketContent` per `Bucket` and vice-versa.
+
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: Bucket
+metadata:
+  name:
+  namespace:
+  labels:
+    cosi.io/provisioner: [1]
+  finalizers:
+  - cosi.io/finalizer [2]
+spec:
+  protocol: [3]
+  bucketPrefix: [4]
+  bucketClassName: [5]
+  accessSecretName: [6]
+status:
+  bucketContentName: [7]
+  phase: [8]
+  conditions: 
+```
+1. `labels`: should be added by controller.  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+1. `finalizers`: should be added by controller to defer `Bucket` deletion until backend deletion ops succeed.
+1. `protocol`: specifies the desired protocol.  One of {“s3”, “gcs”, or “azureBlob”}.
+1. `bucketPrefix`: (Optional) prefix prepended to a randomly generated bucket name, eg. "YosemitePhotos-". If empty no prefix is appended.
+1. `bucketClassName`: Name of the target `BucketClass`.
+1. `userSecretName`: (optional) Secret in the Bucket’s namespace storing credentials to be used by a workload for bucket access.
+1. `bucketContentName`: Name of a bound `BucketContent`.
+1. `phase`: 
+   - *Pending*: The controller has detected the new `Bucket` and begun provisioning operations
+   - *Bound*: Provisioning operations have completed and the `Bucket` has been bound to a `BucketContent`.
+
+#### BucketContent
+
+A cluster-scoped resource representing an object store bucket. The `BucketContent` is expected to store stateful data relevant to bucket deprovisioning. The `BucketContent` is bound to the `Bucket` in a 1:1 mapping. For MVP a `BucketContent` is not reused.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketContent
+Metadata:
+  name: [1]
+  labels:
+    cosi.io/provisioner: [2]
+  finalizers:
+  - cosi.io/finalizer [3]
+spec:
+  provisioner: [4]
+  releasePolicy: [5]
+  accessMode: [6]
+  bucketClassName: [7]
+  bucketRef: [8]
+    name:
+    namespace:
+    uuid:
+  secretRef: [9]
+    name:
+    namespace:
+  protocol: [10]
```

Comment:
Similar to a comment above. I'm not sure this schema should restrict/type the value of this field, and instead would suggest using a `parameters map[string]string` approach.

Reply:
Rationale: a provisioner, even for S3-compatible storage for example, may want to keep more meta-information for whatever reason.

Reply:
There were a couple reasons for this decision (having considered string:string maps early on):

1.  An API that just passes an opaque map to a backend driver is overly generic.  So generic in fact that it no longer adheres to the scope for which it's named.  Drivers could be written to do any number of tasks beyond object storage provisioning through it.  We felt this devalued the API.

2. Strictly defining protocol fields (while providing room for expansions) allows us to deliver connection information in a standard format, rather than an arbitrary format defined by drivers.  The result is that apps won't be coupled to drivers.  IOW, an app moving from an on-prem S3 object store to a cloud S3 object store won't need to be reconfigured to ingest the new BucketContent format, even if the drivers differ.